### PR TITLE
refactor(streaming): remove epoch on state table read interface

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
currently, our streaming executor just read the newest epoch's data from state store, so it is not necessary to handle epoch in every read interface. for the most executors, maintaining the epoch and passing it when read from state table is verbose and unnessary.
If one day we really need it we will add a special interface with epoch and reuse the code in state table. 
TODO: some code clean can be done in hash join and dyn filter
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by hzxa21** [comment]
Do we have streaming executors that won't read from latest? If not, can we always use a MAX_EPOCH (e.g. u64::max) for streaming executor reads to avoid init_epoch?

Thanks for bringing this up :)

**Comment by st1page** [comment]
> Do we have streaming executors that won't read from latest? If not, can we always use a MAX_EPOCH (e.g. u64::max) for streaming executor reads to avoid init_epoch?
> 
> Thanks for bringing this up :)

1. the delta join will read from the previous epoch from the state store(shared buffer), but in that case, it need not to write.
2. I am not sure if we can assume it now. We still have the possibility to add the read interface with epoch on state table.
3. I have think that but a epoch in state table can do more strict check. And in this process, I actually found some executors did not handle the first init epoch properly.

File: src/stream/src/executor/mod.rs [hunk 1]
```diff
@@ -606,6 +608,22 @@ pub async fn expect_first_barrier(
...
+    let message = stream
+        .next()
+        .stack_trace("expect_first_barrier")
+        .await
+        .expect("failed to extract the first message: stream closed unexpectedly")?;
+    let barrier = message
+        .into_barrier()
+        .expect("the first message must be a barrier");
+    Ok(barrier)
+}
```
Comment (st1page) [inline src/stream/src/executor/mod.rs:625]: add a help function to assert the first init barrier on AlignedMessageStream and used in HashJoin and DynFilter

File: src/storage/src/table/streaming_table/state_table.rs [hunk 2]
```diff
@@ -308,6 +313,32 @@ impl<S: StateStore> StateTable<S> {
...
+            None => {
+                self.epoch = Some(epoch);
+            }
+        }
+    }
+
+    /// get the newest epoch of the state store and panic if the `init_epoch()` has never be called
+    pub fn epoch(&self) -> u64 {
+        self.epoch.unwrap_or_else(|| panic!("try to use state table's epoch, but the init_epoch() has not been called, table_id: {}", self.table_id()))
+    }
```
Comment (st1page) [inline src/storage/src/table/streaming_table/state_table.rs:340]: PR's quick index

**Comment by fuyufjh** [comment]
> Do we have streaming executors that won't read from latest? If not, can we always use a MAX_EPOCH (e.g. u64::max) for streaming executor reads to avoid init_epoch?
> 
> Thanks for bringing this up :)

I think `MAX_EPOCH` is actually what we want 🤔

This PR eliminates `epoch` from state_table interface based on this assumption: all reads are performed with the latest epoch. However, it's more straightforward to `MAX_EPOCH` for it, isn't it?

**Comment by wcy-fdu** [comment]
> 1. the delta join will read from the previous epoch from the state store(shared buffer), but in that case, it need not to write. So maybe we can  try not oi use stateTable here

+1 for introducing a `MAX_EPOCH`, and thus maybe we can avoid some complex assert🤔

**Comment by st1page** [comment]
> > 1. the delta join will read from the previous epoch from the state store(shared buffer), but in that case, it need not to write. So maybe we can  try not oi use stateTable here
> 
> +1 for introducing a `MAX_EPOCH`, and thus maybe we can avoid some complex assertthinking

we actually need these assert and I think after the refactor, the stream executor will not need to aware the epoch and there is the only place we can assert the epoch goes back

**Comment by st1page** [comment]
rethink my idea about why I want to maintain the `init_epoch` interface and why I want to store the epoch in the state table. I think I will change the design to store an epoch{prev, curr} in the state table. The reasons are as follows. 
1. **for assertion:** In our checkpoint design, every stream executor must commit this epoch when it receives a barrier. But currently, we do not have a guarantee for that. If an executor does not commit on epoch 1 but commits them on epoch 2 later, everything will just look good before someone really read the data at snapshot epoch 1. Although it seems that no one will forget this when implementing a stream executor, I think it is useful if we can have a reliable guarantee of any streaming executor in troubleshooting.
2. **for most executor's reading:** in this PR, the state table just read from the state store with the "last flushed epoch". That's exactly wrong. if we want to flush an epoch to state store multi times as mentioned in **for writting**. What the executor should read is the `current_epoch`. Furthermore, under our share-nothing-like design, we know that all the data covered by the state table instance will be only written by the state table itself. So we know the current_epoch is the newest epoch and we can use `MAX_EPOCH` when reading from the state store if it can perform better with the "newest epoch" instead of a "specific epoch"
3. **for writing:** currently, the state table will just write once each epoch. But IIRC the state store has supported "flush anytime" which means that state table can flush an epoch to the state store multi times to ensure the memtable is not too big. so we must maintain the current epoch in state table so that it can know the current epoch to write to state store.
4. **for delta join and other fancy executors:** have some offline discussion that we think even in the fanciest situation, the executor will just read the previous epoch snapshot from state store. so I will add an interface to read from the prev epoch.

And I just want to merge this PR first and do the above things later. the PR involves so many files and I have solved so many annoying conflicts. Also, lots of stream executors' refactor is blocked by it. So we can do it later.

**Comment by st1page** [comment]
Another question, can we move state table("src/storage/src/table/streaming_table") into the stream crate? I think it should not used by others.

**Review by hzxa21 — APPROVED** [review]

**Comment by fuyufjh** [comment]
> rethink my idea about why I want to maintain the `init_epoch` interface and why I want to store the epoch in the state table. I think I will change the design to store an epoch{prev, curr} in the state table. The reasons are as follows.
> 
> 1. **for assertion:** In our checkpoint design, every stream executor must commit this epoch when it receives a barrier. But currently, we do not have a guarantee for that. If an executor does not commit on epoch 1 but commits them on epoch 2 later, everything will just look good before someone really read the data at snapshot epoch 1. So if the state table maintain the prev and current epoch, It can assert if the `commit.epoch.prev == self.epoch.curr`. Although it seems that no one will forget this when implementing a stream executor, I think it is useful if we can have a reliable guarantee of any streaming executor in troubleshooting.
> 2. **for most executor's reading:** in this PR, the state table just read from the state store with the "last flushed epoch". That's exactly wrong. if we want to flush an epoch to state store multi times as mentioned in **for writting**. What the executor should read is the `current_epoch`. Furthermore, under our share-nothing-like design, we know that all the data covered by the state table instance will be only written by the state table itself. So we know the current_epoch is the newest epoch and we can use `MAX_EPOCH` when reading from the state store if it can perform better with the "newest epoch" instead of a "specific epoch"
> 3. **for writing:** currently, the state table will just write once each epoch. But IIRC the state store has supported "flush anytime" which means that state table can flush an epoch to the state store multi times to ensure the memtable is not too big. so we must maintain the current epoch in state table so that it can know the current epoch to write to state store.
> 4. **for delta join and other fancy executors:** have some offline discussion that we think even in the fanciest situation, the executor will just read the previous epoch snapshot from state store. so I will add an interface to read from the prev epoch.
> 
> And I just want to merge this PR first and do the above things later. the PR involves so many files and I have solved so many annoying conflicts. Also, lots of stream executors' refactor is blocked by it. So we can do it later. @hzxa21 @fuyufjh @wcy-fdu

I am neutral about it. Please also convince @BugenZhao 😁

**Pull request merged**